### PR TITLE
ASoC: codecs: add support for ES8326

### PR DESCRIPTION
**Pull request by yangxiaohua2009**
The ES8326 codec is not compatible with ES8316 and requires a dedicated driver.

Signed-off-by: David Yang <yangxiaohua@everest-semi.com>

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

**Comment by yangxiaohua2009** [comment]
test this please

**Comment by yangxiaohua2009** [comment]
@plbossart compile issue fixed. The codec driver may need to be rewrited to be compatible with ucm sof-essx8336.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        clocks = <&clks 10>;
+        clock-names = "mclk";
+        #sound-dai-cells = <0>;
+        mic1-src = [22];
+        mic2-src = [44];
+        amic1-src = [19];
+        amic2-src = [19];
+        jack-pol = [0e];
+      };
+    };
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml]: missing newline

Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml]: Usually dt-bindings are in a separate patch (they need to be acked by DT maintainers)

File: sound/soc/codecs/es8326.c [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+{
+	struct es8326_priv *es8326 = dev_id;
+	struct snd_soc_component *comp = es8326->component;
+	unsigned int iface;
+
+	mutex_lock(&es8326->lock);
+	if (!es8326->jack)
+		goto out;
+	snd_soc_component_write(comp, ES8326_ANA_MICBIAS_1B, 0x7c);
+	mdelay(500);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: 500ms delay in an interrupt thread? that sounds wild!

File: sound/soc/codecs/es8326.c [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+	8000, 12000, 16000, 24000, 32000, 48000, 96000,
+};
+
+static struct snd_pcm_hw_constraint_list constraints_12288 = {
+	.count = ARRAY_SIZE(rates_12288),
+	.list = rates_12288,
+};
+
+static unsigned int rates_112896[] = {
+	8000, 11025, 22050, 44100,
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: are you sure about the 8000 here? it's not the same root clock.

File: sound/soc/codecs/es8326.c [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+};
+
+static struct snd_pcm_hw_constraint_list constraints_112896 = {
+	.count = ARRAY_SIZE(rates_112896),
+	.list = rates_112896,
+};
+
+static unsigned int rates_12[] = {
+	8000, 11025, 12000, 16000, 22050, 24000, 32000, 44100, 48000,
+	48000, 88200, 96000,
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: same here, does the 11025 family belong here?

File: sound/soc/codecs/es8326.c [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+}
+
+static void es8326_enable_jack_detect(struct snd_soc_component *component,
+				struct snd_soc_jack *jack)
+{
+	struct es8326_priv *es8326 = snd_soc_component_get_drvdata(component);
+
+	dev_dbg(component->dev, "Enter into %s\n", __func__);
+	es8326->jd_inverted = device_property_read_bool(component->dev,
+							"everest,jack-detect-inverted");
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: no, this needs to be read in the component probe along with all other properties.

File: sound/soc/codecs/es8326.c [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+	enable_irq(es8326->irq);
+	es8326_irq(es8326->irq, es8326);
+}
+
+static void es8326_disable_jack_detect(struct snd_soc_component *component)
+{
+	struct es8326_priv *es8326 = snd_soc_component_get_drvdata(component);
+
+	dev_dbg(component->dev, "Enter into %s\n", __func__);
+	if (!es8326->jack)
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:623]: it's a bit inconsistent, here you are reading the value without using the mutex_lock?

File: sound/soc/codecs/es8326.c [hunk 7]
```diff
@@ -0,0 +1,818 @@
...
+	.num_dapm_routes = ARRAY_SIZE(es8326_dapm_routes),
+	.controls = es8326_snd_controls,
+	.num_controls = ARRAY_SIZE(es8326_snd_controls),
+};
+
+static int es8326_i2c_probe(struct i2c_client *i2c,
+			    const struct i2c_device_id *id)
+{
+	struct es8326_priv *es8326;
+	int ret = -1;
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: this init seems useless?

File: sound/soc/codecs/es8326.c [hunk 8]
```diff
@@ -0,0 +1,818 @@
...
+		.acpi_match_table = ACPI_PTR(es8326_acpi_match),
+		.of_match_table = of_match_ptr(es8326_of_match),
+	},
+	.probe = es8326_i2c_probe,
+};
+module_i2c_driver(es8326_i2c_driver);
+
+MODULE_DESCRIPTION("ASoC es8326 driver");
+MODULE_AUTHOR("David Yang <yangxiaohua@everest-semi.com>");
+MODULE_LICENSE("GPL");
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: newline

**Comment by plbossart** [comment]
test this please

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 9]
```diff
@@ -0,0 +1,110 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/sound/everest,es8326.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Everest ES8326 audio CODEC
+
+maintainers:
+  - Daniel Drake <drake@endlessm.com>
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml]: really? Daniel was not part of this development effort, was he?

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+  amic1-src:
+    description:
+      the value of reg 23 when headset plugged.
+    $ref: /schemas/types.yaml#/definitions/uint8-array
+    minimum: 0x19
+    maximum: 0x29
+    default: 0x19
+
+  amic2-src:
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml]: you need the everest, prefix for all properties?

**Review by plbossart — CHANGES_REQUESTED** [review]
The main issue is the delay in the interrupt thread, and several other cosmetic problems to fix.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml (hunk 1 above)
Comment (yangxiaohua2009) [reply Documentation/devicetree/bindings/sound/everest,es8326.yaml]: Can I submit the dt-binding patch later?

File: sound/soc/codecs/es8326.c (hunk 2 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: A delay is needed for the codec to detect the correct jack state.
Mic bias is turned off to prevent pop sound.
When a jack plugged in, the mic bias is turned on so that the codec can differentiate headphone from headset.
Maybe a delayed_work is preferred?

File: sound/soc/codecs/es8326.c (hunk 5 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: it's the same as es8316.c. sof_es8336 is probed before codec driver though

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml (hunk 1 above)
Comment (plbossart) [reply Documentation/devicetree/bindings/sound/everest,es8326.yaml]: I personally don't care, but keep in mind this will be asked for upstream so you have to work on this.

File: sound/soc/codecs/es8326.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c]: that would sound much better indeed. Most codec drivers use a workqueue for jack detection.

File: sound/soc/codecs/es8326.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c]: > sof_es8336 is probed before codec driver though

that cannot work. deferred probe would be used if that was the case, and the sof_es8336 probe would be restarted at a later point.

**Comment by plbossart** [comment]
test this please

**Comment by plbossart** [comment]
SOFCI TEST

File: sound/soc/codecs/es8326.c [hunk 11]
```diff
@@ -0,0 +1,753 @@
...
+{
+	struct es8326_priv *es8326 = snd_soc_component_get_drvdata(component);
+	int ret;
+
+	es8326->component = component;
+	es8326->mclk = devm_clk_get_optional(component->dev, "mclk");
+	if (IS_ERR(es8326->mclk)) {
+		dev_err(component->dev, "unable to get mclk\n");
+		return PTR_ERR(es8326->mclk);
+	}
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: this is wrong - same mistake on es8316.c 

All devm_ usage has to be done from the I2C device probe.

File: sound/soc/codecs/es8326.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c:623]: i see it's the same in es8316 so probably ok.

File: sound/soc/codecs/es8326.c [hunk 12]
```diff
@@ -0,0 +1,753 @@
...
+
+	es8326->irq = i2c->irq;
+	/* ES8316 is level-based while ES8326 is edge-based */
+	ret = devm_request_threaded_irq(&i2c->dev, es8326->irq, NULL, es8326_irq,
+					IRQF_TRIGGER_RISING | IRQF_ONESHOT,
+					"es8326", es8326);
+	if (ret == 0)
+		disable_irq(es8326->irq);
+	else
+		dev_warn(&i2c->dev, "Getting irq failed.");
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: can this driver work without the IRQ? if no, should an error be returned?

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml (hunk 9 above)
Comment (plbossart) [reply Documentation/devicetree/bindings/sound/everest,es8326.yaml]: ping

**Review by plbossart — CHANGES_REQUESTED** [review]

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml (hunk 9 above)
Comment (yangxiaohua2009) [reply Documentation/devicetree/bindings/sound/everest,es8326.yaml]: removed @plbossart

File: sound/soc/codecs/es8326.c [hunk 13]
```diff
@@ -0,0 +1,754 @@
...
+	u8 amic1_src;
+	u8 amic2_src;
+	u8 mic1_src;
+	u8 mic2_src;
+	u8 jack_pol;
+	bool jd_inverted;
+	unsigned int sysclk;
+};
+
+static void es8326_jack_detect_handler(struct work_struct *work){
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: brace should be at start of line.

File: sound/soc/codecs/es8326.c [hunk 14]
```diff
@@ -0,0 +1,754 @@
...
+};
+MODULE_DEVICE_TABLE(of, es8326_of_match);
+
+static struct i2c_driver es8326_i2c_driver = {
+	.driver = {
+		.name = "es8326",
+		.acpi_match_table = ACPI_PTR(es8326_acpi_match),
+		.of_match_table = of_match_ptr(es8326_of_match),
+	},
+	.probe = es8326_i2c_probe,
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:746]: you need a remove function to cancel the workqueue.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        clock-names = "mclk";
+        #sound-dai-cells = <0>;
+        mic1-src = [22];
+        mic2-src = [44];
+        amic1-src = [19];
+        amic2-src = [19];
+        jack-pol = [0e];
+      };
+    };
+
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml]: this should be a separate patch, so that DT reviewers can take a look without having to look at the driver code.

File: sound/soc/codecs/es8326.c [hunk 16]
```diff
@@ -0,0 +1,754 @@
...
+
+static irqreturn_t es8326_irq(int irq, void *dev_id)
+{
+	struct es8326_priv *es8326 = dev_id;
+	struct snd_soc_component *comp = es8326->component;
+
+	snd_soc_component_write(comp, ES8326_ANA_MICBIAS_1B, 0x7c);
+
+	queue_delayed_work(system_wq, &es8326->jack_detect_work,
+			   msecs_to_jiffies(500));
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:88]: you need to disable the interrupt and cancel the workqueue in a .remove function

**Review by plbossart — CHANGES_REQUESTED** [review]
it's good that you added a workqueue for jack detection, but that needs to be handled in a .remove function.
also please split the DT bindings in a separate patch.

With those changes, I think this is starting to be ready for upstream in the next kernel cycle. Thanks @yangxiaohua2009

File: sound/soc/codecs/es8326.c [hunk 17]
```diff
@@ -0,0 +1,756 @@
...
+#include <linux/module.h>
+#include <sound/jack.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include <sound/soc-dapm.h>
+#include <sound/tlv.h>
+#include "es8326.h"
+
+static int es8326_set_bias_level(struct snd_soc_component *codec,
+				 enum snd_soc_bias_level level);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: this forward declaration doesn't seem required?

File: sound/soc/codecs/es8326.c [hunk 18]
```diff
@@ -0,0 +1,756 @@
...
+
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: it's actually the wrong function, this code doesn't compile at all.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml (hunk 15 above)
Comment (yangxiaohua2009) [reply Documentation/devicetree/bindings/sound/everest,es8326.yaml]: separate patch in #3546

**Comment by plbossart** [comment]
@yangxiaohua2009 This looks good enough to me, please send this patch and the device-tree one to alsa-devel@alsa-project.org and CC: maintainers
Thanks.

**Comment by plbossart** [comment]
closing, this will be handled with an upstream merge.
